src/Compilers/CSharp/Test/Emit3/Semantics/DictionaryExpressionTests.cs

```diff
@@ -1680,27 +1750,35 @@ static void Main()
             }
         }
 
-        [Fact]
-        public void KeyValuePairConversions_Dynamic_02()
+        [Theory]
+        [InlineData(LanguageVersion.CSharp12, "IEnumerable<KeyValuePair<K, V>>")]
+        [InlineData(LanguageVersion.Preview, "IEnumerable<KeyValuePair<K, V>>")]
+        [InlineData(LanguageVersion.Preview, "IDictionary<K, V>")]
+        [InlineData(LanguageVersion.Preview, "Dictionary<K, V>")]
+        public void KeyValuePairConversions_Dynamic_02(LanguageVersion languageVersion, string typeName)
         {
-            string source = """
+            string source = $$"""
                 using System.Collections.Generic;
                 class Program
                 {
-                    static Dictionary<K, V> FromExpression<K, V>(dynamic d) => [d];
-                    static Dictionary<K, V> FromSpread<K, V>(IEnumerable<dynamic> e) => [..e];
+                    static void Main()
+                    {
+                        var x = new KeyValuePair<int, string>(2, "two");
+                        var y = new KeyValuePair<int, string>(3, "three");
+                        FromExpression<int, string>(x).Report();
+                        FromSpread<int, string>(new dynamic[] { y }).Report();
+                    }
+                    static {{typeName}} FromExpression<K, V>(dynamic d) => [d];
+                    static {{typeName}} FromSpread<K, V>(IEnumerable<dynamic> e) => [..e];
                 }
                 """;
-            var comp = CreateCompilation(source);
-            // PROTOTYPE: Should allow implicit conversions from expression and from type for expression element
-            // and spread element, not just implicit conversions between Key and Value types.
-            comp.VerifyEmitDiagnostics(
-                // (4,65): error CS0029: Cannot implicitly convert type 'dynamic' to 'KeyValuePair<K, V>'
-                //     static Dictionary<K, V> FromExpression<K, V>(dynamic d) => [d];
-                Diagnostic(ErrorCode.ERR_NoImplicitConv, "d").WithArguments("dynamic", "System.Collections.Generic.KeyValuePair<K, V>").WithLocation(4, 65),
-                // (5,76): error CS0029: Cannot implicitly convert type 'dynamic' to 'KeyValuePair<K, V>'
-                //     static Dictionary<K, V> FromSpread<K, V>(IEnumerable<dynamic> e) => [..e];
-                Diagnostic(ErrorCode.ERR_NoImplicitConv, "e").WithArguments("dynamic", "System.Collections.Generic.KeyValuePair<K, V>").WithLocation(5, 76));
+            var verifier = CompileAndVerify(
+                [source, s_dictionaryExtensions],
+                parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion),
+                targetFramework: TargetFramework.Net80,
+                verify: Verification.Skipped,
+                expectedOutput: IncludeExpectedOutput("[2:two], [3:three], "));
+            verifier.VerifyDiagnostics();
```

Comment:
Do we have tests verifying the conversions in the semantic model (`GetTypeInfo.Type` or `.ConvertedType` and `GetConversion`)? I worry about the element expression case where we perform part-wise conversion on the key and value. Those part-wise conversions are only store in the assignment.
There's a similar question for how this should be represented in IOperation, but I believe IOperation is already tracked as a whole for follow-up.

Reply:
Added tests, with link to work item.